libsrc/c128/tgi/c128-hi.s

```diff
@@ -0,0 +1,880 @@
+
+X1              := ptr1
+Y1              := ptr2
+X2              := ptr3
+Y2              := ptr4
+TEXT            := ptr3
+
+TEMP            := tmp4
+TEMP2           := sreg
+POINT           := regsave
+
+CHUNK           := X2           ; Used in the line routine
+OLDCHUNK        := X2+1         ; Ditto
+
+; Absolute variables used in the code
+
+.bss
+
+ERROR:          .res    1       ; Error code
+PALETTE:        .res    2       ; The current palette
+
+BITMASK:        .res    1       ; $00 = clear, $FF = set pixels
+
+; Line routine stuff
+DX:             .res    2
+DY:             .res    2
+
+; BAR variables
+X1SAVE:         .res    2
+Y1SAVE:         .res    1
+X2SAVE:         .res    2
+Y2SAVE:         .res    1
+
+; Text output stuff
+TEXTMAGX:       .res    1
+TEXTMAGY:       .res    1
+TEXTDIR:        .res    1
+
+; Constants and tables
+
+.rodata
+
+DEFPALETTE:     .byte   $00, $01        ; White on black
+PALETTESIZE     = * - DEFPALETTE
+
+BITTAB:         .byte   $80,$40,$20,$10,$08,$04,$02,$01
+BITCHUNK:       .byte   $FF,$7F,$3F,$1F,$0F,$07,$03,$01
+
+CHARROM         := $D000                ; Character ROM base address
+
+VBASE           := $C000                ; Video memory base address
+CBASE           := $E000                ; Color memory base address
+
+.code
+
+; ------------------------------------------------------------------------
+; INSTALL routine. Is called after the driver is loaded into memory. May
+; initialize anything that has to be done just once. Is probably empty
+; most of the time.
+;
+; Must set an error code: NO
+;
+
+INSTALL:
+;       rts                     ; fall through
+
+
+; ------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory. May
+; clean up anything done by INSTALL but is probably empty most of the time.
+;
+; Must set an error code: NO
+;
+
+UNINSTALL:
+        rts
+
+
+; ------------------------------------------------------------------------
+; INIT: Changes an already installed device from text mode to graphics
+; mode.
+; Note that INIT/DONE may be called multiple times while the driver
+; is loaded, while INSTALL is only called once, so any code that is needed
+; to initializes variables and so on must go here. Setting palette and
+; clearing the screen is not needed because this is called by the graphics
+; kernel later.
+; The graphics kernel will never call INIT when a graphics mode is already
+; active, so there is no need to protect against that.
+;
+; Must set an error code: YES
+;
+
+INIT:
+
+; Initialize variables.
+
+        ldx     #$FF            ; Foreground color
+        stx     BITMASK
+
+; Switch into graphics mode.
+
+; Select a video bank:
+; bank 0 = $03 ($0000-$3FFF) (default)
+; bank 1 = $02 ($4000-$7FFF)
+; bank 2 = $01 ($8000-$BFFF)
+; bank 3 = $00 ($C000-$FFFF) (TGI)
+
+        lda     CIA2_PRA
+        and     #<~$03          ; Bank 3
```

Comment:
Is it necessary/desirable to explicitly select the low byte here?

Reply:
We get a range error if we don't use `<`.  Ca65 expressions are calculated in the width (32/64 bits) of the `signed long` data type of the compiler that builds ca65.  Currently, the values aren't demoted automatically to the width of 6502 operands -- we must do it with the proper operator or function.

Reply:
Thanks for the explanation :-)